src/field/FieldIntegrator.i.hh

```diff
@@ -0,0 +1,233 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file FieldIntegrator.i.hh
+//---------------------------------------------------------------------------//
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with shared data and the stepper.
+ */
+CELER_FUNCTION
+FieldIntegrator::FieldIntegrator(const FieldParamsPointers& shared,
+                                 RungeKutta&                stepper)
+    : shared_(shared), stepper_(stepper)
+{
+    CELER_ENSURE(shared_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Adaptive step control
+ */
+CELER_FUNCTION
+real_type FieldIntegrator::advance_chord_limited(real_type hstep, ode_type& y)
+{
+    ode_type  yend         = y;
+    real_type curve_length = 0;
+
+    real_type dyerr;
+    real_type h = find_next_chord(hstep, y, yend, dyerr);
+
+    // evaluate a relative error
+    real_type ratio = dyerr / (shared_.epsilon_step * h);
+
+    if (ratio < 1.0)
+    {
+        // Accept this accuracy and update the state
+        y = yend;
+    }
+    else
+    {
+        // Compute the hnext for accuracy_advance
+        real_type hnext = new_step_size(h, ratio);
+
+        // Advance more accurately to the "end of chord"
+        bool is_good = accurate_advance(h, y, curve_length, hnext);
+        if (!is_good)
+        {
+            h = curve_length;
+        }
+    }
+    return h;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * find the next chord and return a step length taken and updates the state
+ */
+CELER_FUNCTION
+real_type FieldIntegrator::find_next_chord(real_type       hstep,
+                                           const ode_type& y,
+                                           ode_type&       yend,
+                                           real_type&      dyerr)
+{
+    // Try with the proposed step
+    real_type h = hstep;
+
+    ode_type dydx;
+    ode_rhs(y, dydx);
+
+    for (CELER_MAYBE_UNUSED int i : celeritas::range(shared_.max_nsteps))
+    {
+        // Always start from the initial point
+        yend = y;
+        real_type dchord_step;
+
+        dyerr = quick_advance(h, yend, dydx, dchord_step);
```

Comment:
Prefix all member function calls with `this->` for improved readability.

Reply:
Done. Also in other classes.